2022Swerve/src/main/java/frc/robot/commands/SwerveTeleopCommand.java

```diff
@@ -0,0 +1,79 @@
+package frc.robot.commands;
+
+import java.util.function.Supplier;
+
+import edu.wpi.first.math.filter.SlewRateLimiter;
+import edu.wpi.first.math.kinematics.ChassisSpeeds;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.Constants.DriveConstants;
+import frc.robot.Constants.OIConstants;
+import frc.robot.subsystems.SwerveSubsystem;
+
+public class SwerveTeleopCommand extends CommandBase {
+
+    private final SwerveSubsystem swerveSubsystem;
+    private final Supplier<Double> xSpeedFunction, ySpeedFunction, turnSpeedFunction;
+    private final Supplier<Boolean> fieldOrientedFunction; //Whether or not control is field oriented
+    private final SlewRateLimiter xLimiter, yLimiter, turnLimiter;
+
+    public SwerveTeleopCommand(SwerveSubsystem swerveSubsystem,
+            Supplier<Double> xSpeedFunction, Supplier<Double> ySpeedFunction, Supplier<Double> turnSpeedFunction,
+            Supplier<Boolean> fieldOrientedFunction) {
+        this.swerveSubsystem = swerveSubsystem;
+        this.xSpeedFunction = xSpeedFunction;
+        this.ySpeedFunction = ySpeedFunction;
+        this.turnSpeedFunction = turnSpeedFunction;
+        this.fieldOrientedFunction = fieldOrientedFunction;
+        this.xLimiter = new SlewRateLimiter(DriveConstants.kMaxAcceleration);
+        this.yLimiter = new SlewRateLimiter(DriveConstants.kMaxAcceleration);
+        this.turnLimiter = new SlewRateLimiter(DriveConstants.kMaxAngluarAcceleration);
+        addRequirements(swerveSubsystem);
+    }
+
+    @Override
+    public void initialize() {}
+
+    @Override
+    public void execute() {
+        //Get joystick inputs
+        double xSpeed = xSpeedFunction.get();
+        double ySpeed = ySpeedFunction.get();
+        double turnSpeed = turnSpeedFunction.get();
+
+        //Apply deadband
+        xSpeed = Math.abs(xSpeed) > OIConstants.kDeadband ? xSpeed : 0.0;
+        ySpeed = Math.abs(ySpeed) > OIConstants.kDeadband ? ySpeed : 0.0;
+        turnSpeed = Math.abs(turnSpeed) > OIConstants.kDeadband ? turnSpeed : 0.0;
+
+        //Smooth input using a slew limiter, map to a speeds in meters/radians per second
+        xSpeed = xLimiter.calculate(xSpeed) * (DriveConstants.kPhysicalMaxSpeed / DriveConstants.kSpeedFactor);
```

Comment:
Limiting the speed for X and Y separately leads to an interesting byproduct.

You can go faster diagonally than you can go either right to left or blue to red.

Draw it out.  Make a circle centered on the origin with X-Axis and Y-Axis.  
You want the radius of that circle to be the maximum speed that the robot can go.
If you go along the positive X axis, your limit will be .25
If you go along the positive Y axis, your limit will be .25
If you go at a 45 degree angle in the first quadrant, then your speed will be sqrt( .25^2 + .25^2 )
Which is .35

Reply:
How to approach this?  Do we care?  

.35 is 1.414 times .25.  So, 40% faster to go diagonally...
Seems important enough to address.  If we wanted to go that fast, we should just go that fast all ways.

You can approach it with trig.  there was also a two dimensional control curve that someone posted to chief delphi in the good/bad swerve control thread.  Might be good to check that out too, and decide if that is a better way to go.

Reply:
I am planning on removing this limiter in the re-implementation of the control system, this PR is primarily just code from the reference, which included the slew limiter.

Reply:
The slew limiter by itself isn't causing this.  This is based on considering X and Y separately.  Do you have a plan for this in the next PR for control system?

Reply:
I do not like the idea of using the speed factor, no. I feel the joysticks should be mapped to the full 100% of physical drivebase capability, and that this factor should be removed.

Reply:
Opened a trello card, consensus was to project all outputs to a unit vector.